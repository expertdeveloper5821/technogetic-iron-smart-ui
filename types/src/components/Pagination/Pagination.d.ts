import './Pagination.css';
type PaginationProps = {
    currentPage: number;
    totalPages: number;
    onPageChange?: (page: number) => void;
};
export declare const Pagination: ({ currentPage, totalPages, onPageChange }: PaginationProps) => JSX.Element;
export {};
