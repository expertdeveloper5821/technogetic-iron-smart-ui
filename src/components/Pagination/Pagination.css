.pagination {
    display: flex;
    padding: 0;
}

.pagination-item {
    margin: 0 2px;
    cursor: pointer;
    padding: 2px 10px;
    display: flex;
    align-items: center;
}

.pagination-item.active {
    font-weight: bold;
    border-radius: 3px;
    color: white;
    background-color: #003554;
}
