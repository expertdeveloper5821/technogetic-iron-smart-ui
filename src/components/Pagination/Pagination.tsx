import React from 'react';
import { NextIcon } from '../../assets/NextIcon';
import { PreviousIcon } from '../../assets/PreviousIcon';
import './Pagination.css';
type PaginationProps = {
    currentPage: number;
    totalPages: number;
    onPageChange: (page: number) => void;
};

export const Pagination = ({ currentPage, totalPages, onPageChange }: PaginationProps) => {
    const isFirstPage = currentPage === 1;
    const isLastPage = currentPage === totalPages;
    const handlePageChange = (page: number) => {
        if (page < 1 || page > totalPages) {
            return;
        }
        onPageChange(page);
    };

    const renderPaginationItem = (page: number, label: string | number) => {
        const isActive = currentPage === page;
        return (
            <div key={page} className={`pagination-item ${isActive ? 'active' : ''}`} onClick={() => handlePageChange(page)}>
                {label}
            </div>
        );
    };

    const renderPaginationItems = () => {
        const paginationItems = [];
        const visiblePages = 5; // Number of visible page links

        // Calculate the start and end page indexes
        let startPage = Math.max(currentPage - Math.floor(visiblePages / 2), 1);
        let endPage = startPage + visiblePages - 1;
        if (endPage > totalPages) {
            endPage = totalPages;
            startPage = Math.max(endPage - visiblePages + 1, 1);
        }

        // Render the previous page link
        paginationItems.push(
            <div key="previous" className={`pagination-item ${isFirstPage ? 'disabled' : ''}`} data-testid="previousIcon" onClick={() => handlePageChange(currentPage - 1)}>
                <PreviousIcon />
            </div>
        );

        // Render the page links
        if (startPage > 1) {
            paginationItems.push(renderPaginationItem(1, 1));
            if (startPage > 2) {
                paginationItems.push(
                    <div key="ellipsis-prev" className="pagination-item ellipsis">
                        ...
                    </div>
                );
            }
        }

        for (let page = startPage; page <= endPage; page++) {
            paginationItems.push(renderPaginationItem(page, page));
        }

        if (endPage < totalPages) {
            if (endPage < totalPages - 1) {
                paginationItems.push(
                    <div key="ellipsis-next" className="pagination-item ellipsis">
                        ...
                    </div>
                );
            }
            paginationItems.push(renderPaginationItem(totalPages, totalPages));
        }

        // Render the next page link
        paginationItems.push(
            <div key="next" className={`pagination-item ${isLastPage ? 'disabled' : ''}`} data-testid="nextIcon" onClick={() => handlePageChange(currentPage + 1)}>
                <NextIcon />
            </div>
        );

        return paginationItems;
    };

    return <div className="pagination">{renderPaginationItems()}</div>;
};
