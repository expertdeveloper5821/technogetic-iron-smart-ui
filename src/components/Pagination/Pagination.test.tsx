import React from 'react';
import { render, fireEvent, screen, getByTestId } from '@testing-library/react';
import { Pagination } from './Pagination';

describe('Pagination Component', () => {
    const onPageChangeMock = jest.fn();
    const totalPages = 10;

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('renders correctly', () => {
        const { container } = render(<Pagination currentPage={1} totalPages={totalPages} onPageChange={onPageChangeMock} />);

        expect(container).toMatchSnapshot();
    });

    test('calls onPageChange with the previous page number', () => {
        render(<Pagination currentPage={2} totalPages={totalPages} onPageChange={onPageChangeMock} />);

        const previousButton = screen.getByTestId('previousIcon');
        fireEvent.click(previousButton);
        expect(onPageChangeMock).toHaveBeenCalledWith(1);
    });

    test('calls onPageChange with the next page number', () => {
        render(<Pagination currentPage={1} totalPages={totalPages} onPageChange={onPageChangeMock} />);

        const nextButton = screen.getByTestId('nextIcon');
        fireEvent.click(nextButton);
        expect(onPageChangeMock).toHaveBeenCalledWith(2);
    });

    test('calls onPageChange with the correct page number when a page link is clicked', () => {
        const { getByText } = render(<Pagination currentPage={3} totalPages={totalPages} onPageChange={onPageChangeMock} />);

        const pageLink = getByText('5');
        fireEvent.click(pageLink);
        expect(onPageChangeMock).toHaveBeenCalledWith(5);
    });
});
