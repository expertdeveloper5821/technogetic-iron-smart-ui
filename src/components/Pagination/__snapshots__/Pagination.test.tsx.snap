// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component renders correctly 1`] = `
<div>
  <div
    class="pagination"
  >
    <div
      class="pagination-item disabled"
      data-testid="previousIcon"
    >
      <svg
        class="icon flat-color"
        data-name="Flat Color"
        fill="#003554"
        height="14"
        id="previous"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.45,2.11a1,1,0,0,0-1.05.09l-12,9a1,1,0,0,0,0,1.6l12,9a1,1,0,0,0,1.05.09A1,1,0,0,0,18,21V3A1,1,0,0,0,17.45,2.11Z"
          id="primary"
          style="fill: #003554;"
        />
      </svg>
    </div>
    <div
      class="pagination-item active"
    >
      1
    </div>
    <div
      class="pagination-item "
    >
      2
    </div>
    <div
      class="pagination-item "
    >
      3
    </div>
    <div
      class="pagination-item "
    >
      4
    </div>
    <div
      class="pagination-item "
    >
      5
    </div>
    <div
      class="pagination-item ellipsis"
    >
      ...
    </div>
    <div
      class="pagination-item "
    >
      10
    </div>
    <div
      class="pagination-item "
      data-testid="nextIcon"
    >
      <svg
        class="icon flat-color"
        data-name="Flat Color"
        fill="#003554"
        height="14"
        id="previous"
        transform="rotate(0)matrix(-1, 0, 0, 1, 0, 0)"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.45,2.11a1,1,0,0,0-1.05.09l-12,9a1,1,0,0,0,0,1.6l12,9a1,1,0,0,0,1.05.09A1,1,0,0,0,18,21V3A1,1,0,0,0,17.45,2.11Z"
          id="primary"
          style="fill: #003554;"
        />
      </svg>
    </div>
  </div>
</div>
`;
